File: Python3/_98.py
"""Binary search tree satisfies 3 conditions:

1. All nodes in left subtree have lower val
2. All nodes in right subtree have higher val
3. Left and right subtrees must also be BST

BST follows inorder (Left-Node-Right) traversal
For inorder, nodes of left subtree are traversed
[before] subtree root and they must have lower val
than root in BST. Likewise, nodes of right subtree
traversed [after] root and they must have higher
val than root in BST
"""

# Iterative
class Solution:
    def isValidBST(self, current) -> bool:
        stack = []
        prev = None
        
        while current or stack:
            if current:
                stack.append(current)
                current = current.left
                
            else:
                current = stack.pop()
                if (prev and prev.val >= current.val):
                    return False
                
                prev = current
                current = current.right
                
        return True

# Recursive
class Solution:
    def isValidBST(self, current):
        return self.BSTUtil(current, float('-inf'), float('inf'))
    
    def BSTUtil(self, current, lo, hi):
        if not current:
            return True
        
        if current.val <= lo:
            return False
        
        if current.val >= hi:
            return False
        
        # For left subtree of current of which current is on right subtree of
        # some ancestor, ensure left subtree is lowerBound by ancestor's val
        # For upper bound of left subtree, use parent/current's

        # For right subtree of current of which current is on left subtree
        # of ancestor, ensure right subtree is upper by ancestor's val
        # For lower bound of right subtree, use parent/current's
        return self.BSTUtil(current.left, lo, current.val) and self.BSTUtil(current.right, current.val, hi)
    
    

File: Cpp/_98.cpp
/*
Binary search tree satisfies 3 conditions:

1. All nodes in left subtree have lower val
2. All nodes in right subtree have higher val
3. Left and right subtrees must also be BST

BST follows inorder (Left-Node-Right) traversal
For inorder, nodes of left subtree are traversed
[before] subtree root and they must have lower val
than root in BST. Likewise, nodes of right subtree
traversed [after] root and they must have higher
val than root in BST
*/

// Iterative
class Solution {
public:
    bool isValidBST(TreeNode* current) {
        stack<TreeNode*> stack;
        TreeNode* prev = nullptr;

        while (current || !stack.empty()) {
            if (current) {
                stack.push(current);
                current = current->left;
            }
            else {
                current = stack.top();
                stack.pop();
                if (prev && prev->val >= current->val) {
                    return false;
                }
                prev = current;
                current = current->right;
            }
        }
        return true;
    }
};

//Recursive

// Passing bound values for recursion as in _98.py returns
// wrong result if any node has val INT_MAX or INT_MIN

// To avoid issues with INT_MAX or INT_MIN comparisons,
// pass nodes providing boundaries directly

class Solution {
public:
    bool isValidBST(TreeNode* current) {
        return BSTUtil(current, nullptr, nullptr);
    }
    
    bool BSTUtil(TreeNode* current, TreeNode* lo, TreeNode* hi) {
        if (!current) return true;
        
        if (lo && current->val <= lo->val) return false;
        if (hi && current->val >= hi->val) return false;
        
        return BSTUtil(current->left, lo, current) && BSTUtil(current->right, current, hi);
    }
};